src/System.Text.CaseFolding/System/Text/CaseFolding/SimpleCaseFolding.cs

```diff
@@ -0,0 +1,606 @@
+            }
+
+            if (source.Length == 0)
+            {
+                destination.Clear();
+                return;
+            }
+
+            // Diagnostics.Assert(destination.Length >= source.Length, "Destination span length must be equal or greater then source span length.");
+            ref char dst = ref MemoryMarshal.GetReference(destination);
+            ref char src = ref MemoryMarshal.GetReference(source);
+
+            var length = source.Length;
+
+            // var l0AsSpan = MapBelow5FF.AsSpan();
+            // ref char refMapBelow5FF = ref MemoryMarshal.GetReference(l0AsSpan);
+            ref char refMapBelow5FF = ref MapBelow5FF[0];
+
+            // For char below 0x5ff use fastest 1-level mapping.
+            while (length != 0 && src <= MaxChar)
+            {
+                dst = Unsafe.Add(ref refMapBelow5FF, src);
+                src = ref Unsafe.Add(ref src, 1);
+                dst = ref Unsafe.Add(ref dst, 1);
+                length--;
+            }
+
+            if (length == 0)
+            {
+                return;
+            }
+
+            ref ushort refMapLevel1 = ref s_MapLevel1;
+            ref char refMapData = ref s_refMapData;
+
+            // We catch a char above 0x5ff.
+            // Process it with more slow two-level mapping.
+            while (length != 0 && !IsSurrogate(src))
+            {
+                var v1 =  Unsafe.Add(ref refMapLevel1, src >> 8);
+                var ch1 = Unsafe.Add(ref refMapData, v1 + (src & 0xFF));
+                if (ch1 == 0)
+                {
+                    ch1 = src;
+                }
+
+                dst = ch1;
+                src = ref Unsafe.Add(ref src, 1);
+                dst = ref Unsafe.Add(ref dst, 1);
+                length--;
+            }
+
+            if (length == 0)
+            {
+                return;
+            }
+
+            ref ushort refMapSurrogateLevel1 = ref s_refMapSurrogateLevel1;
+            ref var refMapSurrogateData = ref s_refMapSurrogateData;
+
+            while (length != 0)
+            {
+                // We catch a high or low surrogate.
+                // Process it and fallback to fastest options.
+                var c1 = src;
+                var isHighSurrogateA = IsHighSurrogate(c1);
+
+                if (isHighSurrogateA)
+                {
+                    // Get low surrogates.
+                    length--;
+                    if (length == 0)
+                    {
+                        // No low surrogate - throw?
+                        throw new ArgumentNullException("Low surrogate is expected.");
+                    }
+
+                    src = ref Unsafe.Add(ref src, 1);
+                    var c1Low = src;
+
+                    if (!IsLowSurrogate(c1Low))
+                    {
+                        // No low surrogate - throw?
+                        throw new ArgumentNullException("Low surrogate is expected.");
```

Comment:
By default Core uses replacement strategy
https://source.dot.net/#System.Private.CoreLib/shared/System/Text/UnicodeEncoding.cs,29
https://source.dot.net/#System.Private.CoreLib/shared/System/Text/Encoding.cs,212
and allow throw strategy.
So I agree to remove throw and implement the replacement with 0xFFFD but we need a conclusion should we implement throw strategy too?
Thinking about targeting the APIs only for comparison/search scenarios I'd conclude that we could implement only replacement strategy because read/write/edit code is this code that should report illegal Unicode.
Can you confirm?

Reply:
You don't even need replacement to 0xFFFD, you are doing case folding, so unpaired surrogate should be treated as any other characters not having case folding and used as ordinal.

---
In reply to: [250551403](https://github.com/dotnet/corefxlab/pull/2637#discussion_r250551403) [](ancestors = 250551403)

Reply:
Today I reviewed Unicode standard - how process illegal Unicode strings. And I don't found a special option for folding. Nearest strategy is "replacement to 0xFFFD". Also there was a reference to security.
Since this code is for searching and comparing, we could just bypass illegal surrogates and get expected results for search/compare.
So my only concern is security but I have no experience.
What will you conclusion?

Reply:
> What will you conclusion?

don't fold the malformed surrogate and treat it as if any other characters that not having case folding.